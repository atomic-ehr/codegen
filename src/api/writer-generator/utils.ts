export const words = (s: string) => {
    return s.split(/(?<=[a-z])(?=[A-Z])|[-_.\s]/).filter(Boolean);
};

export const kebabCase = (s: string) => {
    return words(s)
        .map((s) => s.toLowerCase())
        .join("-");
};

export const capitalCase = (s: string) => {
    if (s.length === 0) throw new Error("Empty string");
    return s[0]?.toUpperCase() + s.substring(1).toLowerCase();
};

export const camelCase = (s: string) => {
    if (s.length === 0) throw new Error("Empty string");
    const [first, ...rest] = words(s);
    return [first?.toLowerCase(), ...rest.map(capitalCase)].join("");
};

export const pascalCase = (s: string) => {
    return words(s).map(capitalCase).join("");
};

export const snakeCase = (s: string) => {
    return words(s)
        .map((s) => s.toLowerCase())
        .join("_");
};

export const uppercaseFirstLetter = (str: string): string => {
    if (!str || str.length === 0) return str;
    return str.charAt(0).toUpperCase() + str.slice(1);
};

export const uppercaseFirstLetterOfEach = (strings: string[]): string[] => {
    return strings.map((str) => uppercaseFirstLetter(str));
};


export function deepEqual<T>(obj1: T, obj2: T): boolean {
  if (obj1 === obj2) return true;

  if (obj1 === null || obj2 === null ||
      typeof obj1 !== 'object' || typeof obj2 !== 'object') {
    return false;
  }

  if (Array.isArray(obj1) && Array.isArray(obj2)) {
    if (obj1.length !== obj2.length) return false;
    return obj1.every((item, index) => deepEqual(item, obj2[index]));
  }

  if (Array.isArray(obj1) || Array.isArray(obj2)) {
    return false;
  }

  const keys1 = Object.keys(obj1) as (keyof T)[];
  const keys2 = Object.keys(obj2) as (keyof T)[];

  if (keys1.length !== keys2.length) return false;

  return keys1.every(key =>
    keys2.includes(key) && deepEqual(obj1[key], obj2[key])
  );
}