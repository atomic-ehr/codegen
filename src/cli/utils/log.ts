/**
 * CLI Utilities - Re-exports from main utils with spinner support
 */

import { createSpinner } from "./spinner";

// Re-export all logger functions from the main utils
export {
  type CodegenLogger,
  complete,
  configure,
  createLogger,
  debug,
  dim,
  error,
  header,
  info,
  type LogOptions,
  list,
  plain,
  progress,
  section,
  step,
  success,
  table,
  warn,
} from "../../utils/codegen-logger";

// Export CLI-specific utilities
export { createSpinner };

/**
 * Run a task with a spinner
 */
export async function withSpinner<T>(
  promise: Promise<T>,
  message: string,
  successMessage?: string,
): Promise<T> {
  const spinner = createSpinner(message).start();
  try {
    const result = await promise;
    spinner.succeed(successMessage || message);
    return result;
  } catch (error) {
    spinner.fail(`Failed: ${message}`);
    throw error;
  }
}
