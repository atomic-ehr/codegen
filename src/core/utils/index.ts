/**
 * Core Utilities Export
 *
 * High-performance utilities for the API-first codegen architecture
 */

// Export moved utilities from lib
export * from "./code";
export * from "./errors";

// Export file utilities
export * from "./file";
export * from "./lib-naming";
export * from "./lib-utils-naming";
export * from "./logger";
// Export naming utilities
export * from "./naming";
// Performance and caching utilities
export { createLRUCache, type LRUCache } from "./performance";
// Export string utilities
export * from "./string";
// Export template utilities
export * from "./template";

/**
 * Utility collections for common operations
 */
export const Utils = {
	// Naming utilities
	naming: {
		toCamelCase: async (str: string) =>
			(await import("./naming")).toCamelCase(str),
		toPascalCase: async (str: string) =>
			(await import("./naming")).toPascalCase(str),
		toSnakeCase: async (str: string) =>
			(await import("./naming")).toSnakeCase(str),
		toKebabCase: async (str: string) =>
			(await import("./naming")).toKebabCase(str),
	},

	// String utilities
	string: {
		truncate: async (str: string, length: number) =>
			(await import("./string")).truncate(str, length),
		slugify: async (str: string) => (await import("./string")).slugify(str),
		similarity: async (str1: string, str2: string) =>
			(await import("./string")).similarity(str1, str2),
	},

	// File utilities
	file: {
		readText: async (path: string) =>
			(await import("./file")).FileReader.readText(path),
		writeText: async (path: string, content: string) =>
			(await import("./file")).FileWriter.writeText(path, content),
		exists: async (path: string) =>
			(await import("./file")).FileReader.exists(path),
	},

	// Template utilities
	template: {
		render: async (template: string, context: any) =>
			(await import("./template")).render(template, context),
		compile: async (template: string) =>
			(await import("./template")).templateEngine.compile(template),
	},
} as const;

/**
 * Batch processing utilities
 */
export const BatchUtils = {
	/**
	 * Process array in batches to avoid memory issues
	 */
	async processBatch<T, R>(
		items: T[],
		processor: (item: T, index: number) => Promise<R> | R,
		options?: {
			batchSize?: number;
			concurrency?: number;
			onProgress?: (completed: number, total: number) => void;
		},
	): Promise<R[]> {
		const { batchSize = 100, concurrency = 5, onProgress } = options || {};
		const results: R[] = [];

		for (let i = 0; i < items.length; i += batchSize) {
			const batch = items.slice(i, i + batchSize);
			const batchPromises = batch.map((item, index) =>
				processor(item, i + index),
			);

			// Process batch with concurrency limit
			const batchResults = await Promise.all(batchPromises);
			results.push(...batchResults);

			if (onProgress) {
				onProgress(Math.min(i + batchSize, items.length), items.length);
			}
		}

		return results;
	},

	/**
	 * Process items with concurrency control
	 */
	async processWithConcurrency<T, R>(
		items: T[],
		processor: (item: T, index: number) => Promise<R>,
		concurrency = 5,
	): Promise<R[]> {
		const results: R[] = new Array(items.length);
		const executing: Promise<void>[] = [];

		for (let i = 0; i < items.length; i++) {
			const promise = processor(items[i], i).then((result) => {
				results[i] = result;
			});

			executing.push(promise);

			if (executing.length >= concurrency) {
				await Promise.race(executing);
				const settled = executing.filter((p, _index) => {
					// Remove settled promises
					return p.then(
						() => false,
						() => false,
					);
				});
				executing.splice(0, executing.length, ...settled);
			}
		}

		await Promise.all(executing);
		return results;
	},
};

/**
 * Memory management utilities
 */
export const MemoryUtils = {
	/**
	 * Get current memory usage
	 */
	getMemoryUsage(): NodeJS.MemoryUsage {
		return process.memoryUsage();
	},

	/**
	 * Format memory size for display
	 */
	formatMemorySize(bytes: number): string {
		const units = ["B", "KB", "MB", "GB"];
		let size = bytes;
		let unitIndex = 0;

		while (size >= 1024 && unitIndex < units.length - 1) {
			size /= 1024;
			unitIndex++;
		}

		return `${size.toFixed(2)} ${units[unitIndex]}`;
	},

	/**
	 * Monitor memory usage
	 */
	createMemoryMonitor(options?: {
		interval?: number;
		threshold?: number;
		onThreshold?: (usage: NodeJS.MemoryUsage) => void;
	}): () => void {
		const {
			interval = 5000,
			threshold = 500 * 1024 * 1024,
			onThreshold,
		} = options || {};

		const intervalId = setInterval(() => {
			const usage = this.getMemoryUsage();

			if (threshold && usage.heapUsed > threshold && onThreshold) {
				onThreshold(usage);
			}
		}, interval);

		return () => clearInterval(intervalId);
	},
};

/**
 * Timing utilities
 */
export const TimingUtils = {
	/**
	 * Measure execution time
	 */
	async time<T>(label: string, operation: () => Promise<T> | T): Promise<T> {
		const start = process.hrtime.bigint();

		try {
			const result = await operation();
			return result;
		} finally {
			const end = process.hrtime.bigint();
			const duration = Number(end - start) / 1_000_000; // Convert to milliseconds
			console.log(`${label}: ${duration.toFixed(2)}ms`);
		}
	},

	/**
	 * Create a timer
	 */
	createTimer(): {
		start: () => void;
		stop: () => number;
		elapsed: () => number;
	} {
		let startTime: bigint | null = null;
		let endTime: bigint | null = null;

		return {
			start: () => {
				startTime = process.hrtime.bigint();
				endTime = null;
			},

			stop: () => {
				if (startTime === null) {
					throw new Error("Timer not started");
				}
				endTime = process.hrtime.bigint();
				return Number(endTime - startTime) / 1_000_000;
			},

			elapsed: () => {
				if (startTime === null) {
					return 0;
				}
				const currentTime = endTime || process.hrtime.bigint();
				return Number(currentTime - startTime) / 1_000_000;
			},
		};
	},

	/**
	 * Debounce function calls
	 */
	debounce<T extends (...args: any[]) => any>(
		func: T,
		wait: number,
	): (...args: Parameters<T>) => void {
		let timeoutId: NodeJS.Timeout | null = null;

		return (...args: Parameters<T>) => {
			if (timeoutId) {
				clearTimeout(timeoutId);
			}

			timeoutId = setTimeout(() => {
				func(...args);
			}, wait);
		};
	},

	/**
	 * Throttle function calls
	 */
	throttle<T extends (...args: any[]) => any>(
		func: T,
		limit: number,
	): (...args: Parameters<T>) => void {
		let inThrottle = false;

		return (...args: Parameters<T>) => {
			if (!inThrottle) {
				func(...args);
				inThrottle = true;
				setTimeout(() => {
					inThrottle = false;
				}, limit);
			}
		};
	},
};

/**
 * Validation utilities
 */
export const ValidationUtils = {
	/**
	 * Check if value is empty
	 */
	isEmpty(value: any): boolean {
		if (value == null) return true;
		if (typeof value === "string") return value.trim().length === 0;
		if (Array.isArray(value)) return value.length === 0;
		if (typeof value === "object") return Object.keys(value).length === 0;
		return false;
	},

	/**
	 * Validate email format
	 */
	isValidEmail(email: string): boolean {
		const emailRegex = /^[^\s@]+@[^\s@]+\.[^\s@]+$/;
		return emailRegex.test(email);
	},

	/**
	 * Validate URL format
	 */
	isValidUrl(url: string): boolean {
		try {
			new URL(url);
			return true;
		} catch {
			return false;
		}
	},

	/**
	 * Validate semantic version
	 */
	isValidSemver(version: string): boolean {
		const semverRegex =
			/^(0|[1-9]\d*)\.(0|[1-9]\d*)\.(0|[1-9]\d*)(?:-((?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*)(?:\.(?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*))*))?(?:\+([0-9a-zA-Z-]+(?:\.[0-9a-zA-Z-]+)*))?$/;
		return semverRegex.test(version);
	},

	/**
	 * Validate JSON string
	 */
	isValidJson(str: string): boolean {
		try {
			JSON.parse(str);
			return true;
		} catch {
			return false;
		}
	},
};

/**
 * Array utilities
 */
export const ArrayUtils = {
	/**
	 * Chunk array into smaller arrays
	 */
	chunk<T>(array: T[], size: number): T[][] {
		const chunks: T[][] = [];
		for (let i = 0; i < array.length; i += size) {
			chunks.push(array.slice(i, i + size));
		}
		return chunks;
	},

	/**
	 * Remove duplicates from array
	 */
	unique<T>(array: T[], keyFn?: (item: T) => any): T[] {
		if (!keyFn) {
			return [...new Set(array)];
		}

		const seen = new Set();
		return array.filter((item) => {
			const key = keyFn(item);
			if (seen.has(key)) {
				return false;
			}
			seen.add(key);
			return true;
		});
	},

	/**
	 * Group array by key
	 */
	groupBy<T, K extends string | number | symbol>(
		array: T[],
		keyFn: (item: T) => K,
	): Record<K, T[]> {
		const groups = {} as Record<K, T[]>;

		for (const item of array) {
			const key = keyFn(item);
			if (!groups[key]) {
				groups[key] = [];
			}
			groups[key].push(item);
		}

		return groups;
	},

	/**
	 * Flatten nested arrays
	 */
	flatten<T>(array: (T | T[])[]): T[] {
		const result: T[] = [];

		for (const item of array) {
			if (Array.isArray(item)) {
				result.push(...this.flatten(item));
			} else {
				result.push(item);
			}
		}

		return result;
	},
};

/**
 * Object utilities
 */
export const ObjectUtils = {
	/**
	 * Deep merge objects
	 */
	merge<T extends Record<string, any>>(target: T, ...sources: Partial<T>[]): T {
		if (!sources.length) return target;

		const source = sources.shift();
		if (!source) return target;

		for (const key in source) {
			const value = source[key];

			if (
				value !== null &&
				typeof value === "object" &&
				!Array.isArray(value)
			) {
				if (!target[key] || typeof target[key] !== "object") {
					target[key] = {} as any;
				}
				this.merge(target[key], value);
			} else {
				target[key] = value as any;
			}
		}

		return this.merge(target, ...sources);
	},

	/**
	 * Deep clone object
	 */
	clone<T>(obj: T): T {
		if (obj === null || typeof obj !== "object") return obj;
		if (obj instanceof Date) return new Date(obj) as any;
		if (Array.isArray(obj)) return obj.map(this.clone) as any;

		const cloned = {} as T;
		for (const key in obj) {
			cloned[key] = this.clone(obj[key]);
		}

		return cloned;
	},

	/**
	 * Pick properties from object
	 */
	pick<T extends Record<string, any>, K extends keyof T>(
		obj: T,
		keys: K[],
	): Pick<T, K> {
		const result = {} as Pick<T, K>;

		for (const key of keys) {
			if (key in obj) {
				result[key] = obj[key];
			}
		}

		return result;
	},

	/**
	 * Omit properties from object
	 */
	omit<T extends Record<string, any>, K extends keyof T>(
		obj: T,
		keys: K[],
	): Omit<T, K> {
		const result = { ...obj } as any;

		for (const key of keys) {
			delete result[key];
		}

		return result;
	},
};
