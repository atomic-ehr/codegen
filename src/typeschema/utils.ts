import type {
    CanonicalUrl,
    ChoiceFieldDeclaration,
    ChoiceFieldInstance,
    Field,
    Identifier,
    ProfileTypeSchema,
    RegularField,
    RegularTypeSchema,
    TypeSchema,
} from "./types";

///////////////////////////////////////////////////////////
// TypeSchema processing

export const groupByPackages = (typeSchemas: TypeSchema[]) => {
    const grouped = {} as Record<string, TypeSchema[]>;
    for (const ts of typeSchemas) {
        const packageName = ts.identifier.package;
        if (!grouped[packageName]) {
            grouped[packageName] = [];
        }
        grouped[packageName].push(ts);
    }
    for (const [_packageName, typeSchemas] of Object.entries(grouped)) {
        typeSchemas.sort((a, b) => a.identifier.name.localeCompare(b.identifier.name));
    }
    return grouped;
};

export const collectComplexTypes = (tss: TypeSchema[]): RegularTypeSchema[] =>
    tss.filter((t) => t.identifier.kind === "complex-type");

export const collectResources = (tss: TypeSchema[]): RegularTypeSchema[] =>
    tss.filter((t) => t.identifier.kind === "resource");

export const collectLogicalModels = (tss: TypeSchema[]): RegularTypeSchema[] =>
    tss.filter((t) => t.identifier.kind === "logical");

export const collectProfiles = (tss: TypeSchema[]) => tss.filter((t) => t.identifier.kind === "profile");

///////////////////////////////////////////////////////////
// Field Processing

export const notChoiceDeclaration = (field: Field): RegularField | ChoiceFieldInstance | undefined => {
    if ((field as ChoiceFieldDeclaration).choices) return undefined;
    if ((field as ChoiceFieldInstance).choiceOf) return field as ChoiceFieldInstance;
    return field as RegularField;
};

///////////////////////////////////////////////////////////
// Type Schema Relations

interface TypeRelation {
    parent: Identifier;
    child: Identifier;
}

const resourceRelatives = (schemas: TypeSchema[]): TypeRelation[] => {
    const regularSchemas = collectResources(schemas);
    const directPairs: TypeRelation[] = [];

    for (const schema of regularSchemas) {
        if (schema.base) {
            directPairs.push({ parent: schema.base, child: schema.identifier });
        }
    }

    const allPairs: TypeRelation[] = [...directPairs];
    const findTransitiveRelatives = (parentRef: Identifier): Identifier[] => {
        const directChildren = directPairs
            .filter((pair) => pair.parent.name === parentRef.name)
            .map((pair) => pair.child);

        const transitiveChildren: Identifier[] = [];
        for (const child of directChildren) {
            transitiveChildren.push(...findTransitiveRelatives(child));
        }

        return [...directChildren, ...transitiveChildren];
    };

    for (const pair of directPairs) {
        const transitiveChildren = findTransitiveRelatives(pair.child);
        for (const transitiveChild of transitiveChildren) {
            if (
                !directPairs.some((dp) => dp.parent.name === pair.parent.name && dp.child.name === transitiveChild.name)
            ) {
                allPairs.push({ parent: pair.parent, child: transitiveChild });
            }
        }
    }

    return allPairs;
};

///////////////////////////////////////////////////////////
// Type Schema Index

type PackageName = string;
export type TypeSchemaIndex = {
    schemaIndex: Record<CanonicalUrl, Record<PackageName, TypeSchema>>;
    relations: TypeRelation[];
    resolve: (id: Identifier) => TypeSchema | undefined;
    resourceChildren: (id: Identifier) => Identifier[];
    hierarchy: (schema: TypeSchema) => TypeSchema[];
    findLastSpecialization: (id: Identifier) => Identifier;
    flatProfile: (schema: ProfileTypeSchema) => ProfileTypeSchema;
};

export const mkTypeSchemaIndex = (schemas: TypeSchema[]): TypeSchemaIndex => {
    const index = {} as Record<CanonicalUrl, Record<PackageName, TypeSchema>>;
    const append = (schema: TypeSchema) => {
        const url = schema.identifier.url;
        if (!index[url]) {
            index[url] = {};
        }
        index[url][schema.identifier.package] = schema;
    };
    for (const schema of schemas) {
        append(schema);
    }
    const relations = resourceRelatives(schemas);

    const resolve = (id: Identifier) => index[id.url]?.[id.package];

    const resourceChildren = (id: Identifier): Identifier[] => {
        return relations.filter((relative) => relative.parent.name === id.name).map((relative) => relative.child);
    };

    const hierarchy = (schema: TypeSchema): TypeSchema[] => {
        const res: TypeSchema[] = [];
        let cur: TypeSchema | undefined = schema;
        while (cur) {
            res.push(cur);
            const base = (cur as RegularTypeSchema).base;
            if (base === undefined) break;
            const resolved = resolve(base);
            if (!resolved) {
                throw new Error(`Failed to resolve base type: ${JSON.stringify(base)}`);
            }
            cur = resolved;
        }
        return res;
    };

    const findLastSpecialization = (id: Identifier): Identifier => {
        const schema = resolve(id);
        if (!schema) return id;

        const nonConstraintSchema = hierarchy(schema).find((s) => s.identifier.kind !== "profile");
        if (!nonConstraintSchema) {
            throw new Error(`No non-constraint schema found in hierarchy for ${id.name}`);
        }
        return nonConstraintSchema.identifier;
    };

    const flatProfile = (schema: ProfileTypeSchema): ProfileTypeSchema => {
        const hierarchySchemas = hierarchy(schema);
        const constraintSchemas = hierarchySchemas.filter((s) => s.identifier.kind === "profile");
        const nonConstraintSchema = hierarchySchemas.find((s) => s.identifier.kind !== "profile");

        if (!nonConstraintSchema)
            throw new Error(`No non-constraint schema found in hierarchy for ${schema.identifier.name}`);

        const mergedFields = {} as Record<string, Field>;
        for (const anySchema of constraintSchemas.slice().reverse()) {
            const schema = anySchema as RegularTypeSchema;
            if (!schema.fields) continue;

            for (const [fieldName, fieldConstraints] of Object.entries(schema.fields)) {
                if (mergedFields[fieldName]) {
                    mergedFields[fieldName] = { ...mergedFields[fieldName], ...fieldConstraints };
                } else {
                    mergedFields[fieldName] = { ...fieldConstraints };
                }
            }
        }

        const deps: { [url: string]: Identifier } = {};
        for (const e of constraintSchemas.flatMap((e) => (e as RegularTypeSchema).dependencies ?? [])) {
            deps[e.url] = e;
        }

        const dependencies = Object.values(deps);

        return {
            ...schema,
            base: nonConstraintSchema.identifier,
            fields: mergedFields,
            dependencies: dependencies,
        };
    };

    return {
        schemaIndex: index,
        relations,
        resolve,
        resourceChildren,
        hierarchy,
        findLastSpecialization,
        flatProfile,
    };
};
