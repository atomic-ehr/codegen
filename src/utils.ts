/**
 * Convert a string into PascalCase.
 * Examples:
 *  - "patient-name" -> "PatientName"
 *  - "Patient name" -> "PatientName"
 *  - "patient_name" -> "PatientName"
 *  - "patientName" -> "PatientName"
 */
export function toPascalCase(input: string): string {
	const parts = input
		.replace(/[^A-Za-z0-9]+/g, " ")
		.split(" ")
		.map((p) => p.trim())
		.filter(Boolean);

	if (parts.length === 0) return "";

	return parts.map((p) => p.charAt(0).toUpperCase() + p.slice(1)).join("");
}

/**
 * Split an array into chunks of a given size.
 *
 * Examples:
 *  - chunkArray([1,2,3,4,5], 2) -> [[1,2],[3,4],[5]]
 *  - chunkArray([], 3) -> []
 *
 * @param arr - The array to split.
 * @param size - The maximum size of each chunk (must be >= 1).
 * @returns An array of chunks (each chunk is an array of T).
 * @throws RangeError if size is less than 1.
 */
export function chunkArray<T>(arr: T[], size: number): T[][] {
	if (!Number.isInteger(size) || size < 1) {
		throw new RangeError("chunk size must be an integer greater than 0");
	}

	const result: T[][] = [];
	if (!arr || arr.length === 0) return result;

	for (let i = 0; i < arr.length; i += size) {
		result.push(arr.slice(i, i + size));
	}

	return result;
}
