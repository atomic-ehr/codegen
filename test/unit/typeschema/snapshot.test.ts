import { describe, expect, it } from "bun:test";
import type { PFS } from "@typeschema-test/utils";
import { fs2ts, r4 } from "@typeschema-test/utils";

describe("Type Schema generation from ValueSet", async () => {
  it.todo("administrative-gender", async () => {
    const fs: PFS = await Bun.file(
      "test/asserts/value-sets/administrative-gender.json",
    ).json();
    const ts = await fs2ts(r4, fs);
    expect(ts).toMatchSnapshot();
  });

  it.todo("all-languages", async () => {
    const fs: PFS = await Bun.file(
      "test/asserts/value-sets/all-languages.json",
    ).json();
    const ts = await fs2ts(r4, fs);
    expect(ts).toMatchSnapshot();
  });

  it.todo("marital-status", async () => {
    const fs: PFS = await Bun.file(
      "test/asserts/value-sets/marital-status.json",
    ).json();
    const ts = await fs2ts(r4, fs);
    expect(ts).toMatchSnapshot();
  });
});

describe("Type Schema generation from FHIR Schema", async () => {
  it.todo("with cardinality", async () => {
    const fs: PFS = await Bun.file(
      "test/asserts/fhir-schemas/with-cardinality.fs.json",
    ).json();
    const ts = await fs2ts(r4, fs);
    expect(ts).toMatchSnapshot();
  });

  it.todo("with resource with string", async () => {
    const fs: PFS = await Bun.file(
      "test/asserts/fhir-schemas/resource-with-string.fs.json",
    ).json();
    const ts = await fs2ts(r4, fs);
    expect(ts).toMatchSnapshot();
  });

  it.todo("with resource with code", async () => {
    const fs: PFS = await Bun.file(
      "test/asserts/fhir-schemas/resource-with-code.fs.json",
    ).json();
    const ts = await fs2ts(r4, fs);
    expect(ts).toMatchSnapshot();
  });

  it.todo("with resource with codable concept", async () => {
    const fs: PFS = await Bun.file(
      "test/asserts/fhir-schemas/resource-with-codable-concept.fs.json",
    ).json();
    const ts = await fs2ts(r4, fs);
    expect(ts).toMatchSnapshot();
  });

  it.todo("with resource with choice", async () => {
    const fs: PFS = await Bun.file(
      "test/asserts/fhir-schemas/resource-with-choice.fs.json",
    ).json();
    const ts = await fs2ts(r4, fs);
    expect(ts).toMatchSnapshot();
  });

  it.todo("with resource with nested type", async () => {
    const fs: PFS = await Bun.file(
      "test/asserts/fhir-schemas/resource-with-nested-type.fs.json",
    ).json();
    const ts = await fs2ts(r4, fs);
    expect(ts).toMatchSnapshot();
  });

  it.todo("with resource with nested type 2", async () => {
    const fs: PFS = await Bun.file(
      "test/asserts/fhir-schemas/resource-with-nested-type-2.fs.json",
    ).json();
    const ts = await fs2ts(r4, fs);
    expect(ts).toMatchSnapshot();
  });

  describe("Custom resource", async () => {
    it.todo("TutorNotification", async () => {
      const fs: PFS = await Bun.file(
        "test/asserts/fhir-schemas/TutorNotification.fs.json",
      ).json();
      const ts = await fs2ts(r4, fs);
      expect(ts).toMatchSnapshot();
    });

    it.todo("TutorNotificationTemplate", async () => {
      const fs: PFS = await Bun.file(
        "test/asserts/fhir-schemas/TutorNotificationTemplate.fs.json",
      ).json();
      const ts = await fs2ts(r4, fs);
      expect(ts).toMatchSnapshot();
    });
  });

  describe("Real world examples", async () => {
    it.todo("coding primitive type", async () => {
      const fs: PFS = await Bun.file(
        "test/asserts/fhir-schemas/coding.fs.json",
      ).json();
      const ts = await fs2ts(r4, fs);
      expect(ts).toMatchSnapshot();
    });

    it.todo("string primitive type", async () => {
      const fs: PFS = await Bun.file(
        "test/asserts/fhir-schemas/string.fs.json",
      ).json();
      const ts = await fs2ts(r4, fs);
      expect(ts).toMatchSnapshot();
    });
  });
});
