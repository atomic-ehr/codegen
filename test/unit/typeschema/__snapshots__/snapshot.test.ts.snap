// Bun Snapshot v1, https://bun.sh/docs/test/snapshots

exports[`FHIR Schema to Type Schema (snapshot) with cardinality 1`] = `
"[
  {
    "identifier": {
      "kind": "complex-type",
      "package": "mypackage",
      "version": "0.0.0",
      "name": "Cardinality",
      "url": "http://hl7.org/fhir/StructureDefinition/Cardinality"
    },
    "base": {
      "kind": "complex-type",
      "package": "hl7.fhir.r4.core",
      "version": "4.0.1",
      "name": "Element",
      "url": "http://hl7.org/fhir/StructureDefinition/Element"
    },
    "fields": {
      "zero_or_one": {
        "type": {
          "kind": "primitive-type",
          "package": "hl7.fhir.r4.core",
          "version": "4.0.1",
          "name": "string",
          "url": "http://hl7.org/fhir/StructureDefinition/string"
        },
        "required": false,
        "excluded": false,
        "array": true,
        "min": 0,
        "max": 1
      },
      "one": {
        "type": {
          "kind": "primitive-type",
          "package": "hl7.fhir.r4.core",
          "version": "4.0.1",
          "name": "string",
          "url": "http://hl7.org/fhir/StructureDefinition/string"
        },
        "required": false,
        "excluded": false,
        "array": true,
        "min": 1,
        "max": 1
      },
      "many": {
        "type": {
          "kind": "primitive-type",
          "package": "hl7.fhir.r4.core",
          "version": "4.0.1",
          "name": "string",
          "url": "http://hl7.org/fhir/StructureDefinition/string"
        },
        "required": false,
        "excluded": false,
        "array": true,
        "min": 42,
        "max": 42
      },
      "one_or_many": {
        "type": {
          "kind": "primitive-type",
          "package": "hl7.fhir.r4.core",
          "version": "4.0.1",
          "name": "string",
          "url": "http://hl7.org/fhir/StructureDefinition/string"
        },
        "required": false,
        "excluded": false,
        "array": true,
        "min": 1
      },
      "zero_or_many": {
        "type": {
          "kind": "primitive-type",
          "package": "hl7.fhir.r4.core",
          "version": "4.0.1",
          "name": "string",
          "url": "http://hl7.org/fhir/StructureDefinition/string"
        },
        "required": false,
        "excluded": false,
        "array": true
      }
    },
    "description": "Base StructureDefinition for Coding Type: A reference to a code defined by a terminology system.",
    "dependencies": [
      {
        "kind": "complex-type",
        "package": "hl7.fhir.r4.core",
        "version": "4.0.1",
        "name": "Element",
        "url": "http://hl7.org/fhir/StructureDefinition/Element"
      },
      {
        "kind": "primitive-type",
        "package": "hl7.fhir.r4.core",
        "version": "4.0.1",
        "name": "string",
        "url": "http://hl7.org/fhir/StructureDefinition/string"
      }
    ]
  }
]"
`;

exports[`FHIR Schema to Type Schema (snapshot) with resource with string 1`] = `
"[
  {
    "identifier": {
      "kind": "resource",
      "package": "mypackage",
      "version": "0.0.0",
      "name": "WithPrimitiveString",
      "url": "http://example.io/fhir/WithPrimitiveString"
    },
    "base": {
      "kind": "resource",
      "package": "hl7.fhir.r4.core",
      "version": "4.0.1",
      "name": "DomainResource",
      "url": "http://hl7.org/fhir/StructureDefinition/DomainResource"
    },
    "fields": {
      "someString": {
        "type": {
          "kind": "primitive-type",
          "package": "hl7.fhir.r4.core",
          "version": "4.0.1",
          "name": "string",
          "url": "http://hl7.org/fhir/StructureDefinition/string"
        },
        "required": true,
        "excluded": false,
        "array": false
      }
    },
    "description": "description",
    "dependencies": [
      {
        "kind": "resource",
        "package": "hl7.fhir.r4.core",
        "version": "4.0.1",
        "name": "DomainResource",
        "url": "http://hl7.org/fhir/StructureDefinition/DomainResource"
      },
      {
        "kind": "primitive-type",
        "package": "hl7.fhir.r4.core",
        "version": "4.0.1",
        "name": "string",
        "url": "http://hl7.org/fhir/StructureDefinition/string"
      }
    ]
  }
]"
`;

exports[`FHIR Schema to Type Schema (snapshot) with resource with code 1`] = `
"[
  {
    "identifier": {
      "kind": "resource",
      "package": "mypackage",
      "version": "0.0.0",
      "name": "WithCode",
      "url": "http://example.io/fhir/WithCode"
    },
    "base": {
      "kind": "resource",
      "package": "hl7.fhir.r4.core",
      "version": "4.0.1",
      "name": "DomainResource",
      "url": "http://hl7.org/fhir/StructureDefinition/DomainResource"
    },
    "fields": {
      "gender": {
        "type": {
          "kind": "primitive-type",
          "package": "hl7.fhir.r4.core",
          "version": "4.0.1",
          "name": "code",
          "url": "http://hl7.org/fhir/StructureDefinition/code"
        },
        "required": false,
        "excluded": false,
        "array": false,
        "binding": {
          "kind": "binding",
          "package": "shared",
          "version": "1.0.0",
          "name": "AdministrativeGender",
          "url": "urn:fhir:binding:AdministrativeGender"
        },
        "enum": [
          "male",
          "female",
          "other",
          "unknown"
        ]
      }
    },
    "description": "description",
    "dependencies": [
      {
        "kind": "binding",
        "package": "shared",
        "version": "1.0.0",
        "name": "AdministrativeGender",
        "url": "urn:fhir:binding:AdministrativeGender"
      },
      {
        "kind": "primitive-type",
        "package": "hl7.fhir.r4.core",
        "version": "4.0.1",
        "name": "code",
        "url": "http://hl7.org/fhir/StructureDefinition/code"
      },
      {
        "kind": "resource",
        "package": "hl7.fhir.r4.core",
        "version": "4.0.1",
        "name": "DomainResource",
        "url": "http://hl7.org/fhir/StructureDefinition/DomainResource"
      }
    ]
  },
  {
    "identifier": {
      "kind": "binding",
      "package": "shared",
      "version": "1.0.0",
      "name": "AdministrativeGender",
      "url": "urn:fhir:binding:AdministrativeGender"
    },
    "type": {
      "kind": "primitive-type",
      "package": "hl7.fhir.r4.core",
      "version": "4.0.1",
      "name": "code",
      "url": "http://hl7.org/fhir/StructureDefinition/code"
    },
    "valueset": {
      "kind": "value-set",
      "package": "hl7.fhir.r4.core",
      "version": "4.0.1",
      "name": "AdministrativeGender",
      "url": "http://hl7.org/fhir/ValueSet/administrative-gender"
    },
    "strength": "required",
    "enum": [
      "male",
      "female",
      "other",
      "unknown"
    ],
    "dependencies": [
      {
        "kind": "primitive-type",
        "package": "hl7.fhir.r4.core",
        "version": "4.0.1",
        "name": "code",
        "url": "http://hl7.org/fhir/StructureDefinition/code"
      },
      {
        "kind": "value-set",
        "package": "hl7.fhir.r4.core",
        "version": "4.0.1",
        "name": "AdministrativeGender",
        "url": "http://hl7.org/fhir/ValueSet/administrative-gender"
      }
    ]
  }
]"
`;
